### PR TITLE
[ConstraintSystem] Add locator path elements for pack expansion patterns and pack shapes.

### PR DESCRIPTION
**Pull request by hborla**
Attach these locators to type variables created for pack expansion patterns and pack shapes, as well as a few other pack-specific constraints. Note that `matchTypes` already appends two `PackType` path elements before calling `matchPackTypes`, which covers the other FIXME comments about adding specific locators.

### CODE REVIEW

File: lib/AST/ASTPrinter.cpp [hunk 1]
```diff
@@ -6407,8 +6407,6 @@ class TypePrinter : public TypeVisitor<TypePrinter> {
   void visitElementArchetypeType(ElementArchetypeType *T) {
     if (Options.PrintForSIL) {
       Printer << "@element(\"" << T->getOpenedElementID() << ") ";
-      visit(T->getGenericEnvironment()->getOpenedExistentialType());
```
Comment (slavapestov) [inline lib/AST/ASTPrinter.cpp:6410]: Oops, my bad :)

File: lib/AST/ParameterPack.cpp [hunk 2]
```diff
@@ -314,15 +314,36 @@ CanPackType PackType::getReducedShape() {
...
   if (auto *packType = getAs<PackType>())
     return packType->getReducedShape();
 
   if (auto *expansionType = getAs<PackExpansionType>())
     return expansionType->getReducedShape();
 
+  struct PatternTypeWalker : public TypeWalker {
+    CanType shapeType;
+
+    Action walkToTypePre(Type type) override {
```
Comment (slavapestov) [inline lib/AST/ParameterPack.cpp:326]: The PatternTypeWalker looks like it's the same thing as calling TypeBase::getTypeParameterPacks() and then getting the reduced shape of one of the results? However now I see a bug in TypeBase::getTypeParameterPacks():

```
      if (archetypeTy->isRoot())
        typeParams.insert(paramTy);
```

if your pattern only references a non-root archetype, its going to skip it. We should include non-root archetypes in the result too. This is called from resolvePackExpansionType() in TypeCheckType.cpp among other places, which means we missed test cases involving associated types of type parameter packs

**Review by slavapestov — APPROVED** [review]

**Review by xedin — APPROVED** [review]
Constraint system related changes look good to me!

File: lib/AST/ParameterPack.cpp (hunk 2 above)
Comment (hborla) [reply lib/AST/ParameterPack.cpp:326]: Ah, you're right, thanks! I totally forgot about `getTypeParameterPacks`. I'll fix the issue with pack references via associated types

**Comment by hborla** [comment]
@swift-ci please smoke test

**Review by slavapestov — APPROVED** [review]

**Pull request merged**